### PR TITLE
RFC: improve show_stack() performance

### PR DESCRIPTION
**Pull request by kevinyamauchi**
edit 7.27.18: updated to include prototypes for show_spots (see 5525d9b, 48ba1e7, dd941e2)

### Overview
The show_stack tool is really useful for evaluating the images during processing. However, it is very slow and thus difficult to use. I have prototyped a faster way to update images (see [the notebook](https://github.com/spacetx/starfish/blob/ky-show-stack-perf/notebooks/Faster_show_stack.ipynb)) and proposed a method for speeding up the show_spots as well (see below). Thoughts?

### Approach
**Faster scrolling**
The current version in master creates a new imshow() plot each time the slider is adjusted. Recreating the entire plot is very slow, so instead we use the imshow().set_data() to update just the image data. We should be able to make show_stack(). If we need to go faster, we may need to look into a different plotting library (e.g., pyqtgraph).

**Faster spots**
To speed up the spot display, we draw all of the spots on the image (as if they were z projected) before calling interact object. We set all spots to set_visible = False (i.e., make them invisible). Then, when the viewer "scrolls" between slices, toggle the visible property for spots that are members of that frame. We precompute the masks for each slice to speed up the upate function.

**Image query cursor tooltop**
Additionally, note that I use the 'notebook' (as opposed to 'inline') matplotlib magic, which both allows the draw to work and as a bonus gives an image tooltip. The tooltip is super useful because it reads out the coordinates and intensity of the pixel under the cursor (exactly what we were talking about today)!

### Questions

- Will the proposed show_spots method generalize to other slicing (i.e., not along z)?
- Is the show_spots responsive enough? We could consider blitting (i.e., only interacting with spots that change state between slices), but it may not be worth the effort.
- What class will be passed to show_spots as the 'results_df'?
- Does it matter what backend (e.g., QtAgg or TkAgg) the user has?

### CODE REVIEW

**Comment by ambrosejcarr** [comment]
> Ah sorry about the failed tests - I'll try to figure out the notebook creation later today.

Not a problem. 
```
pip3 install nbencdec


> @ambrosejcarr, I have a question about the results_df that gets passed into show_spots() argument: under the new IntensityTable/Codebook, what class will be passed to show_spots?

That's a great question. I think it would probably be an `IntensityTable`, but I'll have to update the code to make it work. Referencing #324 Because I think that there isn't any coverage of that method right now, and adding tests will help.

**Comment by ttung** [comment]
Generally, you should run `pip install -r REQUIREMENTS-DEV.TXT -r REQUIREMENTS-NOTEBOOK.TXT`

**Comment by ambrosejcarr** [comment]
I was able to load up the notebook -- I love the `%matplotlib notebook`, the tools are super useful; they are indeed exactly what we wanted! 

Unfortunately, when I use that tool and try normal matplotlib figure plots, they don't show up anymore. I haven't done any due diligence yet to see if there are any easy solutions, but we should make sure it doesn't disrupt the regular plotting. 

I tried replacing it with `%matplotlib inline`, but the new plot doesn't survive the transition. It also tried using both inline and notebook, but it seems like the notebook only supports one at a time.

**Comment by kevinyamauchi** [comment]
Thanks for looking into the plotting, Ambrose. I took a look and we can plot using matplotlib, but when we use the %matplotlib notebook magic, we need to explicitly assign the plot object to a fig/axis (see below. plt.plot() just returns the plot object. I also added some examples to [the notebook](https://github.com/spacetx/starfish/blob/ky-show-stack-perf/notebooks/Faster_show_stack.ipynb).

``` python
fig = plt.figure();
ax = fig.add_subplot(111);
ax.plot(range(10));
```

**Comment by ambrosejcarr** [comment]
Got it. Thanks for looking into that. I think the requirement that we have to assign to axes is an acceptable price to pay to get pan/zoom and better performance on the show stack. 

Interested to see what @dganguli has to say about this. :)

**Comment by kevinyamauchi** [comment]
I agree. I also think that the specificity of assigning the plot to an axis is useful for when we are generating multiple plots (e.g., QC plots) and want to manipulate a particular one (either statically or dynamically).

**Comment by ambrosejcarr** [comment]
I like these changes, here's what I think it would take to merge them: 

1. Since `%matplotlib notebook` causes plots not to show without explicit figure references, and the proposed code updates breaks show_stack in `%matplotlib inline`, we'll need to update existing notebooks to use the proposed strategy. 
2. In updated notebooks, verify that new `show_spots` has logical results
3. [Optional] Add some kind of tests to the vis methods
4. Remove example notebook

@kevinyamauchi Do you feel comfortable working through these? We can give you some pointers on how we use `nbencdec` to generate `.py` files that pair with the notebooks if you'd like. Alternatively, I'd be happy to pair on this with you (or perhaps @ttung could in person).

**Comment by kevinyamauchi** [comment]
Sounds good. I can work through those items. I think I will need some help/discussion for creating tests (item 3) and generating the py files for notebooks, but I can take a crack at it myself first.

With regards to the spot data input format for show_spots(), should I think I will integrate the input format in the current implementation just to get it tested quickly with the current notebooks. We can work to add support for IntensityTable soon. How does that sound?

**Comment by ambrosejcarr** [comment]
> Sounds good. I can work through those items. I think I will need some help/discussion for creating tests (item 3) and generating the py files for notebooks, but I can take a crack at it myself first

Cool, I don't have much experience with vis tests, so I'm more interested than anything else. If it seems too complicated I wouldn't be too upset if you punted on it (that's what we're currently doing, see #324)

> With regards to the spot data input format for show_spots(), should I think I will integrate the input format in the current implementation just to get it tested quickly with the current notebooks. We can work to add support for IntensityTable soon. How does that sound?

Sounds like a good plan.

File: starfish/image/_stack.py [hunk 1]
```diff
@@ -335,25 +335,58 @@ def show_stack(
 
         show_spot_function = self._show_spots
 
+        # Create the plot
+        fig, ax = plt.subplots(figsize=figure_size)
+        im = ax.imshow(linear_view[0], cmap=color_map)
+        ax.set_xticks([])
+        ax.set_yticks([])
+
+        if show_spots:
```
[the remark this answers is not in the document]
Comment (dganguli) [reply starfish/image/_stack.py]: alternatively, you could continue to use _show_spots, but pass in a spots table that looks like this:

```results = pd.DataFrame({'y':intensities.x, 'x':intensities.y, 'r':intensities.r})
f, ax = plt.subplots(figsize=(20, 20))
ax.imshow(blobs, cmap=plt.cm.gray)

stack._show_spots(results, ax=plt.gca())```

File: starfish/image/_stack.py [hunk 2]
```diff
@@ -335,25 +335,58 @@ def show_stack(
 
         show_spot_function = self._show_spots
```
[the remark this answers is not in the document]
Comment (kevinyamauchi) [reply starfish/image/_stack.py]: Sorry - deleted.

File: starfish/image/_stack.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevinyamauchi) [reply starfish/image/_stack.py]: We discussed this on slack, but just adding it here to consolidate the record. I just had this in as a placeholder until I figured out the expected format for the spot data. I have no implemented show_spots for the pandas data frame as you suggested here.

**Comment by kevinyamauchi** [comment]
Sounds good, Deep! I fixed show_spots to work as a drop in for the current notebooks. I tested the notebooks and they all seem to work with minor tweaks. I can go over this with you when we meet tomorrow. If things look okay to you, I'll take a shot at making the notebook .py files.

@ambrosejcarr, the one thing I can't get to work is pixel spot detector in the MERFISH notebook. I get a "KeyError: 'barcode'". Is this something being addressed in a different PR (or perhaps already merged with master)? However, this doesn't seem related to the show_stack() so I am not sure we need to fix it in this PR.

``` python
from starfish.pipeline.features.pixels.pixel_spot_detector import PixelSpotDetector
psd = PixelSpotDetector(
    codebook='https://s3.amazonaws.com/czi.starfish.data.public/MERFISH/codebook.csv',
    distance_threshold=0.5176,
    magnitude_threshold=1,
    area_threshold=2,
    crop_size=40
)

spot_attributes, decoded = psd.find(s)
```

File: notebooks/Faster_show_stack.ipynb [hunk 3]
```diff
@@ -0,0 +1,3401 @@
+{
```
Comment (dganguli) [inline notebooks/Faster_show_stack.ipynb]: i say we nuke this notebook.

**Comment by dganguli** [comment]
and one last thing -- we should verify the new notebooks work -- which i'm pretty sure they weill.

**Comment by ambrosejcarr** [comment]
I like the extra expressiveness of: 

```python
f, ax = plt.subplots(figsize=(n, n))
``` 
This is just another "pro" for me. :)

**Comment by kevinyamauchi** [comment]
Okay, I finally got the notebooks updated and encoded as .py files (thanks, @ttung for the help!). The allen notebook still fails due to the experiment.json file needing updates. Since that is a separate issue, if everyone is okay with the changes, can we merge? Thoughts @dganguli and @ambrosejcarr ?

**Pull request merged**